hack/scripts/provision.sh

```diff
@@ -0,0 +1,996 @@
+	install_tar "$url" "$(dirname $INSTALL_PATH)" || die "could not install containerd"
+
+	"$CONTAINERD_BIN" --version &>/dev/null
+	ok_or_die "Containerd version $tag not installed"
+
+	say "Containerd version $tag successfully installed"
+}
+
+# Prepare the containerd state dirs
+make_containerd_dirs() {
+	local dirs=("$DEVMAPPER_DIR" "$CONTAINERD_STATE_DIR" "$(dirname $CONTAINERD_CONFIG_PATH)")
+	for d in "${dirs[@]}"; do
+		say "Creating containerd directory $d"
+
+		mkdir -p "$d" || die "Failed to make containerd directory $d"
+	done
+
+	say "All containerd directories created"
+}
+
+# Write out the containerd config file
+write_containerd_config() {
+	local thinpool="$1"
+
+	say "Writing containerd config to $CONTAINERD_CONFIG_PATH"
+
+	cat <<EOF >"$CONTAINERD_CONFIG_PATH"
+version = 2
+
+root = "$CONTAINERD_ROOT_DIR"
+state = "$CONTAINERD_STATE_DIR"
+
+[grpc]
+  address = "$CONTAINERD_STATE_DIR/containerd.sock"
+
+[metrics]
+  address = "127.0.0.1:1338"
+
+[plugins]
+  [plugins."io.containerd.snapshotter.v1.devmapper"]
+    pool_name = "$thinpool-thinpool"
+    root_path = "$DEVMAPPER_DIR"
+    base_image_size = "10GB"
+    discard_blocks = true
+
+[debug]
+  level = "trace"
+EOF
+
+	say "Containerd config saved"
+}
+
+# Start the containerd systemd service
+start_containerd_service() {
+	say "Starting containerd service"
+
+	service=$(basename "$CONTAINERD_SERVICE_FILE")
+	fetch_service_file "$CONTAINERD_REPO" "$service" "$CONTAINERD_SERVICE_FILE"
+
+	sed -i "s|ExecStart=.*|& --config $CONTAINERD_CONFIG_PATH|" "$CONTAINERD_SERVICE_FILE"
+
+	start_service "$CONTAINERD_BIN"
+
+	say "Containerd running"
+}
+
+## DIRECT_LVM
+#
+#
+do_all_direct_lvm() {
+	local disk="$1"
+	local thinpool="$2"
+
+	say "Setting up direct_lvm thinpool $thinpool"
+
+	if [[ -z "$disk" ]]; then
+		say_warn "WARNING: -d/--disk has not been set. If you continue, the" \
+			"script will attempt to detect a free disk for formatting. Any data" \
+			"will be lost."
+		get_user_confirmation "Are you sure you wish to continue? (y/n) " || die "Aborted."
+
+		disk=$(find_free_disk || die "Could not detect free disk")
+	fi
+
+	local disk_path="/dev/$disk"
+
+	say "Will use $disk_path for direct-lvm thinpool $thinpool"
+	say_warn "All existing data on $disk_path will be overwritten."
+	get_user_confirmation || die "Aborted."
+
+	create_physical_volume "$disk_path"
+	create_volume_group "$disk_path" "$thinpool"
+	create_logical_volume "$thinpool"
+	apply_lvm_profile "$thinpool"
+	monitor_lvm_profile "$thinpool" || die "failed to monitor lvm profile"
+
+	say "Thinpool $thinpool-thinpool is ready for use"
+}
+
+# Naively find a space block device which is not in use
+# This is really unsafe as it only looks at lsblk to decide anything
+# It is a much better idea to use the --disk flag and pass in something you
+# know will be available
+find_free_disk() {
+	disks=("$(lsblk -o NAME,TYPE,MOUNTPOINT | awk '/disk/ {print $1}')")
```

Comment:
Should we check if it's mounted already?

Reply:
done